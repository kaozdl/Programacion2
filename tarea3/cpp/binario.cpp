/* 4580203 - 4660754 */
#include "../include/utils.h"
#include "../include/texto.h"
#include "../include/info.h"
#include "../include/lista.h"
#include "../include/binario.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>

struct rep_binario {
  info_t dato;
  rep_binario *izq;
  rep_binario *der;
};
nat max (const nat a, const nat b){
  if (a < b) return b;
  else return a;
}

bool Es_Hoja(const binario b){
  return ((b->der == NULL) && (b->izq == NULL));
}

binario crear_binario() {
  binario b = NULL;
  return b;
}

bool insertar_en_binario(const info_t i, binario &b) {
    if (b==NULL){
        b= new rep_binario;
        b->dato =i;
        b->izq=NULL;
        b->der=NULL;
        return true;
    }
    else if (comparar_texto(texto_info(i),texto_info(b->dato))==mayor)
             return insertar_en_binario(i,b->der);
    else if (comparar_texto(texto_info(i),texto_info(b->dato))==menor)
             return insertar_en_binario(i,b->izq);
    else
        return false;
  }


info_t remover_mayor(binario &b) {
  info_t res;
  if (b->der == NULL) {
    res = b->dato;
    binario izq = b->izq;
    delete (b);
    b = izq;
  } else {
    res = remover_mayor(b->der);
  }
  return res;
}

binario crear_filtrado(const int clave, comp_t criterio, const binario b) {
  if (b == NULL)
    return NULL;
  else {
    binario fizq = crear_filtrado(clave,criterio,b->izq);
    binario fder = crear_filtrado(clave,criterio,b->der);
    if (criterio == menor){
      if (clave > numero_info(b->dato)){
        binario fil = new rep_binario;
        fil->dato = crear_info(numero_info(b->dato), copiar_texto(texto_info(b->dato)));
        fil->izq = fizq;
        fil->der = fder;
        return fil;
      }
      else if (es_vacio_binario(fizq)){
        return fder;
      }
      else if (es_vacio_binario(fder)){
        return fizq;
      }
      else {
        info_t mayor = remover_mayor(fizq);
        binario fil = new rep_binario;
        fil->dato = crear_info(numero_info(mayor), copiar_texto(texto_info(mayor)));
        fil->izq = fizq;
        fil->der = fder;
        return fil;
      }
    }
    else if (criterio == mayor){
      if (clave < numero_info(b->dato)){
        binario fil = new rep_binario;
        fil->dato = crear_info(numero_info(b->dato), copiar_texto(texto_info(b->dato)));
        fil->izq = fizq;
        fil->der = fder;
        return fil;
      }
      else if (es_vacio_binario(fizq)){
        return fder;
      }
      else if (es_vacio_binario(fder)){
        return fizq;
      }
      else {
        info_t mayor = remover_mayor(fizq);
        binario fil = new rep_binario;
        fil->dato = crear_info(numero_info(mayor), copiar_texto(texto_info(mayor)));
        fil->izq = fizq;
        fil->der = fder;
        return fil;
      }
    }
    else {
      if (clave == numero_info(b->dato)){
        binario fil = new rep_binario;
        fil->dato = crear_info(numero_info(b->dato), copiar_texto(texto_info(b->dato)));
        fil->izq = fizq;
        fil->der = fder;
        return fil;
      }
      else if (es_vacio_binario(fizq)){
        return fder;
      }
      else if (es_vacio_binario(fder)){
        return fizq;
      }
      else {
        info_t mayor = remover_mayor(fizq);
        binario fil = new rep_binario;
        fil->dato = crear_info(numero_info(mayor), copiar_texto(texto_info(mayor)));
        fil->izq = fizq;
        fil->der = fder;
        return fil;
      }
    }

  }

}




void remover_de_binario(const texto_t t, binario &b) {
  if (b != NULL){
    if (comparar_texto(t,texto_info(b->dato))==menor)
      remover_de_binario(t,b->izq);
    if (comparar_texto(t,texto_info(b->dato))==mayor)
      remover_de_binario(t,b->der);
    if (comparar_texto(t,texto_info(b->dato))==igual){
      if (b->izq == NULL){
        binario borrar = b;
        b = b->der;
        delete borrar;
      }
     else if (b->der == NULL){
       binario borrar = b;
       b = b->izq;
       delete borrar;
     }
     else{

       b->dato=remover_mayor(b->izq);
     }
    }
  }
}

void liberar_binario(binario &b) {
  if (b!=NULL){
    liberar_binario(b->izq);
    liberar_binario(b->der);
    liberar_info(b->dato);
    }
}

bool es_vacio_binario(const binario b) {
  return b == NULL;
}

info_t raiz_binario(const binario b) {
  return b->dato;
}

binario izquierdo(const binario b) {
  return b->izq;
}

binario derecho(const binario b) {
  return b->der;
}

binario buscar_subarbol(const texto_t t, const binario b) {
  if (b == NULL)
    return b;
  else{
    if (comparar_texto(texto_info(b->dato),t)==menor)
      return buscar_subarbol(t,b->der);
    else if (comparar_texto(texto_info(b->dato),t)==mayor)
      return buscar_subarbol(t,b->izq);
    else
      return b;
  }
}

nat altura_binario(const binario b) {
  if (es_vacio_binario(b))
   return 0;
  else
   return (1+max(altura_binario(b->izq),altura_binario(b->der)));
  }


nat cantidad_binario(const binario b) {
  if (b == NULL)
    return 0;
  else
    return 1 + cantidad_binario(b->izq) + cantidad_binario(b->der);
}

nat cantidad_de_caminos_aux(localizador loc, const lista l, const binario b){
  if (!es_localizador_lista(loc) || es_vacio_binario(b))  {
    return 0;
  }
  else if (numero_info(info_lista(loc,l)) != numero_info(b->dato))
    return 0;
      else{
    if(Es_Hoja(b) && es_final_lista(loc,l))
      return 1;
    else{
      loc = siguiente(loc,l);
      return cantidad_de_caminos_aux(loc,l,derecho(b))+cantidad_de_caminos_aux(loc,l,izquierdo(b));
    }
  }

}

nat cantidad_de_caminos(const lista l, const binario b) {
  localizador aux = inicio_lista(l);
  nat cantidad = cantidad_de_caminos_aux(aux,l,b);
  return cantidad;
}

/*
  Auxiliar para camino_a_texto.
 */
static void agregar_rama(char *cadena, rama_t r) {
  switch (r) {
  case izq:
    strcat(cadena, "I");
    break;
  case der:
    strcat(cadena, "D");
    break;
  }
}

texto_t camino_a_texto(const camino_t c) {
  texto_t res;
  // espacio suficiente para "No existe" o para cada rama y la coma y la
  // cantidad, el guión y los paréntesis.
  char cadena[c.cantidad_ramas * 2 + 10];
  if (!c.existe) {
    strcpy(cadena, "No existe");
  } else {
    sprintf(cadena, "%d", c.cantidad_ramas);
    strcat(cadena, "-(");
    if (c.cantidad_ramas > 0) {
      agregar_rama(cadena, c.ramas[0]);
      for (nat i = 1; i < c.cantidad_ramas; i++) {
        strcat(cadena, ",");
        agregar_rama(cadena, c.ramas[i]);
      }
    }
    strcat(cadena, ")");
  }
  res = cadena_a_texto(cadena);
  return res;
}

void buscar_camino_aux(const lista l,localizador loc ,const binario b , camino_t & camino) {
    if (!es_localizador_lista(loc)||es_vacio_binario(b))
        camino.existe=false;
    else if (comparar_texto(texto_info(info_lista(loc,l)),texto_info(b->dato))!=igual)
        camino.existe = false;
    else{
      localizador sig = siguiente(loc,l);
      if (b->izq!=NULL && b->der!=NULL) {
      camino.ramas[camino.cantidad_ramas]=der;
      camino.cantidad_ramas++;
      camino.existe = true;
      buscar_camino_aux(l,sig,derecho(b),camino);
      if (!camino.existe){
        camino.ramas[camino.cantidad_ramas-1]=izq;
        camino.existe = true;
        buscar_camino_aux(l,sig,izquierdo(b),camino);
      }
    }
      else if (b->izq==NULL && b->der!=NULL) {
        camino.ramas[camino.cantidad_ramas]=der;
        camino.cantidad_ramas++;
        camino.existe = true;
        buscar_camino_aux(l,sig,derecho(b),camino);
        if (!camino.existe){
          camino.ramas[camino.cantidad_ramas-1]=izq;
          camino.existe = true;
          buscar_camino_aux(l,sig,izquierdo(b),camino);
        }
      }
      else if (b->izq!=NULL && b->der==NULL) {
        camino.ramas[camino.cantidad_ramas]=izq;
        camino.cantidad_ramas++;
        camino.existe = true;
        buscar_camino_aux(l,sig,izquierdo(b),camino);
        if (!camino.existe){
        camino.ramas[camino.cantidad_ramas-1]=der;
        camino.existe = true;
        buscar_camino_aux(l,sig,derecho(b),camino);
      }
    }
    else if (b->izq==NULL && b->der!=NULL) {
      camino.ramas[camino.cantidad_ramas]=der;
      camino.cantidad_ramas++;
      camino.existe = true;
      buscar_camino_aux(l,sig,derecho(b),camino);
      if (!camino.existe){
        camino.ramas[camino.cantidad_ramas-1]=izq;
        camino.existe = true;
        buscar_camino_aux(l,sig,izquierdo(b),camino);
      }
    }
    else{
        if (es_localizador_lista(sig))
            camino.existe=false;
    }
    }
}
camino_t buscar_camino(const lista l, const binario b) {
        camino_t camino;
        camino.existe = true;
        camino.ramas = new rama_t[altura_binario(b)-1];
        camino.cantidad_ramas=0;
        localizador loc = inicio_lista(l);
        buscar_camino_aux(l,loc,b,camino);
        return camino;
}


void imprimir_binario_auxiliar (const nat k, const binario b){
  if (b != NULL) {
    imprimir_binario_auxiliar(k+1, b->der);
    for (nat i = 0; i < k; i++)
      printf("-");
    escribir_texto(info_a_texto(b->dato));
    printf("\n");
    imprimir_binario_auxiliar (k+1,b->izq);
  }
}

void imprimir_binario(const binario b) {
if (b != NULL){
  int k = 0;
  imprimir_binario_auxiliar(k,b);
  }
}
