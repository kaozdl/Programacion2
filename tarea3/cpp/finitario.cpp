/* 4580203 - 4660754 */
#include "../include/texto.h"
#include "../include/info.h"
#include "../include/lista.h"
#include "../include/uso_lista_arboles.h"
#include "../include/finitario.h"

#include <stdlib.h>

struct rep_finitario {
  info_t dato;
  rep_finitario *primer_hijo;
  rep_finitario *sig_hermano;
};

//AUXILIARES DE FINITARIO
//Devuelve el Maximo entre dos numeros naturales.
nat max2 (const nat a, const nat b){
  if (a < b) return b;
  else return a;
}
//Devuelve true si el finitario f es hoja y false en caso contrario.
bool es_hoja (const finitario f) {
  if ((f->primer_hijo == NULL) && (f->sig_hermano == NULL)) return true;
  else return false;
}


finitario crear_finitario() {
  finitario fin = NULL;
  return fin;
}

void insertar_en_finitario(const info_t i, const nat k, finitario &f) {
  finitario insertar = new rep_finitario;
    insertar->dato=i;
    insertar->primer_hijo=NULL;
    if (f==NULL){
        f=insertar;
        f->sig_hermano=NULL;
    }

    else if (f->primer_hijo==NULL && k!= 1) {
        f->primer_hijo=insertar;
        insertar->sig_hermano=NULL;
    }
    else {
        finitario resto = f->primer_hijo;

        for (nat i = 1; i<k-1 && resto->sig_hermano!=NULL;i++){
           resto = resto->sig_hermano;

        }
        if (k!=1){
        insertar->sig_hermano=resto->sig_hermano;
        resto->sig_hermano=insertar;
      }
       else {
       insertar->sig_hermano=f->primer_hijo;
       f->primer_hijo=insertar;

       }
}
}

void liberar_finitario(finitario &f) {
  if (f!=NULL){
       liberar_finitario(f->primer_hijo);
       liberar_finitario(f->sig_hermano);
       liberar_info(f->dato);
       delete f;
       f=NULL;
    }
}

bool es_vacio_finitario(const finitario f) {
  return (f == NULL);
}

info_t raiz_finitario(const finitario f) {
  return f->dato;
}

finitario hijo(const nat k, const finitario f) {
    if (k==1)
        return f->primer_hijo;
    else if (f->primer_hijo!=NULL){
        finitario son = f->primer_hijo;
        for (nat i = 1; i < k && son!=NULL; i++) {
          son = son->sig_hermano;
        }

        return son;
    }
    else return NULL;
  }

void nivel_aux(lista &lst, const nat k, const finitario f){
  if (f != NULL){
    localizador loc1 = final_lista(lst);
    if (k == 1){
      info_t info = crear_info(numero_info(f->dato), copiar_texto(texto_info(f->dato)));
      insertar_despues(info,loc1,lst);
      loc1 = final_lista(lst);
      nivel_aux(lst,k,f->sig_hermano);
    }
    else {
      nivel_aux(lst,k-1,f->primer_hijo);
      nivel_aux(lst,k,f->sig_hermano);

    }
  }
}

lista nivel(nat k, const finitario f) {
  lista niv = crear_lista();
  nivel_aux(niv,k,f);
  ordenar(niv);
  return niv;
}

nat altura_finitario(const finitario f) {
  if (es_vacio_finitario(f)) return 0;
  else
    return (max2(1+altura_finitario(f->primer_hijo),(altura_finitario(f->sig_hermano))));
  }
