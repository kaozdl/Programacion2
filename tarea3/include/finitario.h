/*
  Módulo de definición de `finitario`.

  Se definen los árboles finitarios de elementos de tipo `info_t`.
  Los hijos de cada nodo se enumeran de izquierda a derecha. El hijo de más a
  la izquierda es el primero.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
 */

#ifndef _FINITARIO_H
#define _FINITARIO_H

#include "utils.h"
#include "info.h"
#include "lista.h"

struct rep_finitario; // representación de `finitario`
typedef rep_finitario *finitario;

/* Constructoras */

/*
  Devuelve un finitario vacío (sin elementos).
 */
finitario crear_finitario();

/*
  Inserta `i` en `f`.
  Si es_vacio_finitario(f), `f` pasa a ser un árbol hoja con elemento `i`.
  En otro caso, `i` pasa a ser el `k`-ésimo hijo de `f`, o el último si la
  cantidad de hijos de `f` es menor que `k` - 1. Los otros hijos de `f`
  mantienen su orden relativo.
  Precondición: `k > 0`.
 */
void insertar_en_finitario(const info_t i, const nat k, finitario &f);

/* Destructoras */

/*
  Libera la memoria asociada a `f` y todos sus elementos.
 */
void liberar_finitario(finitario &f);

/* Predicados */

/*
  Devuelve `true` si `f`es vacío (no tiene elementos), o `false` en otro caso.
 */
bool es_vacio_finitario(const finitario f);

/* Selectoras */

/*
  Devuelve el elemento asociado a la raiz de `f`.
  Precondición: ! es_vacio_finitario(f).
 */
info_t raiz_finitario(const finitario f);

/*
  Devuelve el subárbol con raíz en el `k`-ésimo hijo `f`.
  Si `k` es mayor que la cantidad de hijos de `f` devuelve el árbol vacío.
  Precondiciones: ! es_vacio_finitario(f)` y `k > 0`.
 */
finitario hijo(const nat k, const finitario f);

/*
  Devuelve una lista de elementos de tipo `info_t` de los nodos que están en el
  nivel `k` de `f`. La lista debe estar ordenada de manera creciente según el
  dato numérico. Para cada par de elementos con el mismo dato numérico debe
  quedar primero el que en `f` está más a la izquierda.
  La raíz está en el nivel 1.
  Devuelve la lista vacía si el árbol no tiene nodos en el nivel k.
  La lista devuelta no comparte memoria con `f`.
  Precondición: k >= 1.
 */
lista nivel(nat k, const finitario f);

/*
  Devuelve la altura de `f`.
  La altura de una árbol finitario vacío es 0.
 */
nat altura_finitario(const finitario f);

#endif
