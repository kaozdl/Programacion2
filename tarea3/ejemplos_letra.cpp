/*
  Módulo principal ejemplos_letra.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
*/

#include "include/utils.h"
#include "include/texto.h"
#include "include/info.h"
#include "include/lista.h"
#include "include/binario.h"
#include "include/finitario.h"
#include "include/uso_lista_arboles.h"

#include <stdio.h>

int main() {

  binario b;
  finitario f, h;
  lista lst;
  info_t info;
  texto_t txt;
  bool res;
  camino_t cam;

  // insertar_en_binario
  printf("insertar_en_binario. \n");
  // primer ejemplo
  printf("1:\n");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_en_binario(info, b);
  imprimir_binario(b);
  liberar_binario(b);

  // segundo ejemplo
  printf("2:\n");
  b = crear_binario();
  info = crear_info(8, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_en_binario(info, b);
  imprimir_binario(b);

  // tercer ejemplo
  printf("3:\n");
  info = crear_info(5, cadena_a_texto("d"));
  res = insertar_en_binario(info, b);
  imprimir_binario(b);
  liberar_binario(b);
  if (!res)
    liberar_info(info);

  // remover_de_binario
  printf("remover_de_binario. \n");
  // primer ejemplo
  printf("1:\n");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("k"));
  insertar_en_binario(info, b);
  txt = cadena_a_texto("j");
  remover_de_binario(txt, b);
  imprimir_binario(b);
  liberar_binario(b);
  liberar_texto(txt);

  // segundo ejemplo
  printf("2:\n");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("k"));
  insertar_en_binario(info, b);
  info = crear_info(2, cadena_a_texto("m"));
  insertar_en_binario(info, b);
  txt = cadena_a_texto("m");
  remover_de_binario(txt, b);
  imprimir_binario(b);
  liberar_binario(b);
  liberar_texto(txt);

  // tercer ejemplo
  printf("3:\n");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("k"));
  insertar_en_binario(info, b);
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  info = crear_info(5, cadena_a_texto("x"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("l"));
  insertar_en_binario(info, b);
  info = crear_info(1, cadena_a_texto("n"));
  insertar_en_binario(info, b);
  info = crear_info(3, cadena_a_texto("m"));
  insertar_en_binario(info, b);
  info = crear_info(7, cadena_a_texto("z"));
  insertar_en_binario(info, b);
  txt = cadena_a_texto("x");
  remover_de_binario(txt, b);
  imprimir_binario(b);
  liberar_binario(b);
  liberar_texto(txt);

  // altura_binario
  printf("altura_binario. \n");
  // primer ejemplo
  printf("1:");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  printf("%d \n", altura_binario(b));
  liberar_binario(b);

  // segundo ejemplo
  printf("2:");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  info = crear_info(5, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  printf("%d \n", altura_binario(b));
  liberar_binario(b);

  // tercer ejemplo
  printf("3:");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  info = crear_info(5, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_binario(info, b);
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_binario(info, b);
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  printf("%d \n", altura_binario(b));
  liberar_binario(b);

  // buscar_subarbol
  printf("buscar_subarbol. \n");
  // primer ejemplo
  printf("1:\n");
  b = crear_binario();
  info = crear_info(4, cadena_a_texto("g"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_binario(info, b);
  info = crear_info(7, cadena_a_texto("k"));
  insertar_en_binario(info, b);
  info = crear_info(3, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  info = crear_info(1, cadena_a_texto("e"));
  insertar_en_binario(info, b);
  info = crear_info(8, cadena_a_texto("i"));
  insertar_en_binario(info, b);
  info = crear_info(6, cadena_a_texto("p"));
  insertar_en_binario(info, b);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  info = crear_info(8, cadena_a_texto("h"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("m"));
  insertar_en_binario(info, b);

  txt = cadena_a_texto("i");
  binario sub = buscar_subarbol(txt, b);
  imprimir_binario(sub);
  liberar_binario(b);
  liberar_texto(txt);

  // imprimir_binario
  printf("imprimir_binario. \n");
  // primer ejemplo
  printf("1:\n");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  imprimir_binario(b);
  liberar_binario(b);

  // segundo ejemplo
  printf("2:\n");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  info = crear_info(5, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  imprimir_binario(b);
  liberar_binario(b);

  // tercer ejemplo
  printf("3:\n");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  info = crear_info(5, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_binario(info, b);
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_binario(info, b);
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  imprimir_binario(b);
  liberar_binario(b);

  // cantidad_de_caminos
  printf("cantidad_de_caminos. \n");
  lst = crear_lista();
  info = crear_info(3, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);

  // primer ejemplo
  printf("1:");
  b = crear_binario();
  info = crear_info(3, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  printf("%d \n", cantidad_de_caminos(lst, b));

  // segundo ejemplo
  printf("2:");
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  printf("%d \n", cantidad_de_caminos(lst, b));

  // tercer ejemplo
  printf("3:");
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  printf("%d \n", cantidad_de_caminos(lst, b));

  // cuarto ejemplo
  printf("4:");
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_binario(info, b);
  printf("%d \n", cantidad_de_caminos(lst, b));

  liberar_binario(b);
  liberar_lista(lst);

  // buscar_camino
  printf("buscar_camino. \n");
  lst = crear_lista();
  info = crear_info(4, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);

  // primer ejemplo
  printf("1:");
  b = crear_binario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  cam = buscar_camino(lst, b);
  txt = camino_a_texto(cam);
  escribir_texto(txt);
  escribir_nueva_linea();
  delete[] cam.ramas;

  // segundo ejemplo
  printf("2:");
  info = crear_info(5, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  cam = buscar_camino(lst, b);
  txt = camino_a_texto(cam);
  escribir_texto(txt);
  escribir_nueva_linea();
  delete[] cam.ramas;

  // tercer ejemplo
  printf("3:");
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_binario(info, b);
  cam = buscar_camino(lst, b);
  txt = camino_a_texto(cam);
  escribir_texto(txt);
  escribir_nueva_linea();
  delete[] cam.ramas;

  // cuarto ejemplo
  printf("4:");
  info = crear_info(3, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_binario(info, b);
  cam = buscar_camino(lst, b);
  txt = camino_a_texto(cam);
  escribir_texto(txt);
  escribir_nueva_linea();
  delete[] cam.ramas;

  liberar_binario(b);
  liberar_lista(lst);

  // filtrar_binario
  printf("filtrar_binario. \n");
  // primer ejemplo
  printf("1:\n");
  b = crear_binario();
  info = crear_info(5, cadena_a_texto("g"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_binario(info, b);
  info = crear_info(7, cadena_a_texto("k"));
  insertar_en_binario(info, b);
  info = crear_info(3, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  info = crear_info(1, cadena_a_texto("e"));
  insertar_en_binario(info, b);
  info = crear_info(8, cadena_a_texto("i"));
  insertar_en_binario(info, b);
  info = crear_info(6, cadena_a_texto("p"));
  insertar_en_binario(info, b);
  info = crear_info(9, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  info = crear_info(8, cadena_a_texto("h"));
  insertar_en_binario(info, b);
  info = crear_info(4, cadena_a_texto("m"));
  insertar_en_binario(info, b);

  binario flt = crear_filtrado(5, menor, b);
  imprimir_binario(flt);
  liberar_binario(b);
  liberar_binario(flt);

  // insertar_en_finitario
  printf("insertar_en_finitario. \n");
  f = crear_finitario();

  // primer ejemplo
  printf("1:");
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_finitario(info, 1, f);
  imprimir_finitario(f);

  // segundo ejemplo
  printf("2:");
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_finitario(info, 2, f);
  imprimir_finitario(f);

  // tercer ejemplo
  printf("3:");
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_finitario(info, 1, f);
  imprimir_finitario(f);

  // cuarto ejemplo
  printf("4:");
  info = crear_info(1, cadena_a_texto("v"));
  insertar_en_finitario(info, 4, f);
  imprimir_finitario(f);

  liberar_finitario(f);

  // hijo

  printf("hijo. \n");
  // primer ejemplo
  printf("1:");
  f = crear_finitario();
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_finitario(info, 1, f);
  h = hijo(1, f);
  imprimir_finitario(h);
  liberar_finitario(f);

  // segundo ejemplo
  printf("2:");
  f = crear_finitario();
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_finitario(info, 1, f);
  info = crear_info(7, cadena_a_texto("c"));
  insertar_en_finitario(info, 1, f);
  h = hijo(2, f);
  imprimir_finitario(h);
  liberar_finitario(f);

  // tercer ejemplo
  printf("3:");
  f = crear_finitario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_finitario(info, 1, f);
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_finitario(info, 1, f);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_finitario(info, 1, f);
  h = hijo(1, f);
  info = crear_info(7, cadena_a_texto("z"));
  insertar_en_finitario(info, 1, h);
  h = hijo(2, f);
  info = crear_info(1, cadena_a_texto("v"));
  insertar_en_finitario(info, 1, h);
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_finitario(info, 1, h);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_finitario(info, 1, h);

  h = hijo(2, f);
  imprimir_finitario(h);

  // cuarto ejemplo
  printf("4:");
  finitario h2 = hijo(3, h);
  imprimir_finitario(h2);

  liberar_finitario(f);

  // nivel
  f = crear_finitario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_finitario(info, 1, f);
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_finitario(info, 1, f);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_finitario(info, 1, f);

  h = hijo(1, f);
  info = crear_info(7, cadena_a_texto("z"));
  insertar_en_finitario(info, 1, h);

  h = hijo(2, f);
  info = crear_info(1, cadena_a_texto("v"));
  insertar_en_finitario(info, 1, h);
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_finitario(info, 1, h);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_finitario(info, 1, h);

  printf("nivel. \n");
  // primer ejemplo
  printf("1:");
  lst = nivel(1, f);
  imprimir_lista(lst);
  liberar_lista(lst);

  // segundo ejemplo
  printf("2:");
  lst = nivel(2, f);
  imprimir_lista(lst);
  liberar_lista(lst);

  // tercer ejemplo
  printf("3:");
  lst = nivel(3, f);
  imprimir_lista(lst);
  liberar_lista(lst);

  liberar_finitario(f);

  // kesimo_subarbol
  printf("kesimo_subarbol. \n");
  // primer ejemplo
  printf("1:\n");
  b = crear_binario();
  info = crear_info(15, cadena_a_texto("b"));
  insertar_en_binario(info, b);
  sub = kesimo_subarbol(1, b);
  imprimir_binario(sub);

  // segundo ejemplo
  printf("2:\n");
  info = crear_info(-1, cadena_a_texto("d"));
  insertar_en_binario(info, b);
  sub = kesimo_subarbol(2, b);
  imprimir_binario(sub);

  // tercer ejemplo
  printf("3:\n");
  info = crear_info(34, cadena_a_texto("a"));
  insertar_en_binario(info, b);
  info = crear_info(10, cadena_a_texto("c"));
  insertar_en_binario(info, b);
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_binario(info, b);
  sub = kesimo_subarbol(4, b);
  imprimir_binario(sub);

  liberar_binario(b);

  // imprimir_finitario
  printf("imprime_finitario. \n");
  // primer ejemplo
  printf("1:\n");

  f = crear_finitario();
  info = crear_info(1, cadena_a_texto("b"));
  insertar_en_finitario(info, 1, f);
  info = crear_info(2, cadena_a_texto("a"));
  insertar_en_finitario(info, 1, f);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_en_finitario(info, 1, f);
  h = hijo(1, f);
  info = crear_info(7, cadena_a_texto("z"));
  insertar_en_finitario(info, 1, h);
  h = hijo(2, f);
  info = crear_info(1, cadena_a_texto("v"));
  insertar_en_finitario(info, 1, h);
  info = crear_info(7, cadena_a_texto("t"));
  insertar_en_finitario(info, 1, h);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_en_finitario(info, 1, h);

  imprimir_finitario(f);
  liberar_finitario(f);

  return 0;
}
