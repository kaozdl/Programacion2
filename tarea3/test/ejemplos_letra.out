insertar_en_binario. 
1:
(1,u)
2:
--(3,t)
-(2,d)
(8,a)
3:
--(3,t)
-(2,d)
(8,a)
remover_de_binario. 
1:
(1,k)
2:
(1,k)
3:
--(7,z)
-(1,n)
---(3,m)
--(4,l)
(1,k)
-(2,a)
altura_binario. 
1:1 
2:2 
3:3 
buscar_subarbol. 
1:
(8,i)
-(8,h)
imprimir_binario. 
1:
(1,b)
2:
-(5,d)
(1,b)
3:
--(7,t)
-(5,d)
--(4,c)
(1,b)
-(2,a)
cantidad_de_caminos. 
1:0 
2:1 
3:2 
4:1 
buscar_camino. 
1:0-()
2:No existe
3:No existe
4:2-(D,I)
filtrar_binario. 
1:
-(4,m)
(1,e)
--(2,d)
-(4,c)
--(3,b)
insertar_en_finitario. 
1:(2,a)
2:(2,a)
(7,t)
3:(2,a)
(4,c)(7,t)
4:(2,a)
(1,v)(4,c)(7,t)
hijo. 
1:
2:
3:(2,a)
(1,v)(4,c)(7,t)
4:(1,v)
nivel. 
1:(1,b)
2:(2,d)(2,a)
3:(1,v)(4,c)(7,z)(7,t)
kesimo_subarbol. 
1:
(15,b)
2:
(-1,d)
3:
-(7,t)
(-1,d)
-(10,c)
imprime_finitario. 
1:
(1,b)
(2,d)(2,a)
(1,v)(4,c)(7,z)(7,t)
